File: Assignment-4/java/ScheduleManagerGUI.java
import javax.swing.*;
import javax.swing.border.EmptyBorder;
import javax.swing.table.DefaultTableModel;
import java.awt.*;
import java.awt.event.*;
import java.util.*;

/**
 * GUI for Employee Schedule Manager (Java Swing).
 * Input: employee names and shift preferences per day (entered by user).
 * Output: final weekly schedule in a readable format.
 */
public class ScheduleManagerGUI extends JFrame {

    private static final String[] DAYS = {
        "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"
    };
    private static final String[] SHIFTS = { "morning", "afternoon", "evening" };

    private final DefaultTableModel tableModel;
    private final JTable employeeTable;
    private final JTextArea outputArea;
    private final JLabel statusLabel;
    private final java.util.List<ScheduleManager.Employee> employees = new ArrayList<>();

    public ScheduleManagerGUI() {
        setTitle("Employee Schedule Manager");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setSize(820, 620);
        setLocationRelativeTo(null);

        JPanel main = new JPanel(new BorderLayout(10, 10));
        main.setBorder(new EmptyBorder(12, 12, 12, 12));

        // Top: buttons
        JPanel top = new JPanel(new FlowLayout(FlowLayout.LEFT));
        top.add(new JLabel("Add employees and set their shift preferences, then generate the schedule."));
        JButton addBtn = new JButton("Add employee");
        addBtn.addActionListener(e -> openAddEmployeeDialog());
        top.add(addBtn);
        JButton removeBtn = new JButton("Remove selected");
        removeBtn.addActionListener(e -> removeSelected());
        top.add(removeBtn);
        JButton sampleBtn = new JButton("Load sample data");
        sampleBtn.addActionListener(e -> loadSampleData());
        top.add(sampleBtn);
        JButton clearBtn = new JButton("Clear all");
        clearBtn.addActionListener(e -> clearAll());
        top.add(clearBtn);
        JButton generateBtn = new JButton("Generate schedule");
        generateBtn.addActionListener(e -> generateSchedule());
        top.add(generateBtn);
        statusLabel = new JLabel(" ");
        top.add(statusLabel);
        main.add(top, BorderLayout.NORTH);

        // Center: employee list (name + summary) and output
        JPanel center = new JPanel(new BorderLayout(5, 5));
        String[] cols = { "Name", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun" };
        tableModel = new DefaultTableModel(cols, 0) {
            @Override
            public boolean isCellEditable(int row, int column) { return false; }
        };
        employeeTable = new JTable(tableModel);
        employeeTable.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
        employeeTable.getTableHeader().setReorderingAllowed(false);
        employeeTable.setRowHeight(22);
        center.add(new JScrollPane(employeeTable), BorderLayout.CENTER);
        JPanel centerSouth = new JPanel(new BorderLayout());
        centerSouth.setBorder(BorderFactory.createTitledBorder("Final schedule"));
        outputArea = new JTextArea(16, 72);
        outputArea.setEditable(false);
        outputArea.setFont(new Font(Font.MONOSPACED, Font.PLAIN, 12));
        outputArea.setLineWrap(false);
        outputArea.setText("Add employees (name + preferred shift per day), then click 'Generate schedule'.");
        centerSouth.add(new JScrollPane(outputArea), BorderLayout.CENTER);
        center.add(centerSouth, BorderLayout.SOUTH);
        main.add(center, BorderLayout.CENTER);

        setContentPane(main);
    }

    private void openAddEmployeeDialog() {
        AddEmployeeDialog dlg = new AddEmployeeDialog(this);
        dlg.setVisible(true);
        if (dlg.getResult() != null) {
            addEmployeeToTable(dlg.getResult());
        }
    }

    private void addEmployeeToTable(ScheduleManager.Employee emp) {
        employees.add(emp);
        String[] prefs = new String[7];
        for (int d = 0; d < 7; d++) {
            String p = emp.preferences != null && emp.preferences.containsKey(d)
                ? emp.preferences.get(d) : "-";
            prefs[d] = p.substring(0, 1).toUpperCase();
        }
        tableModel.addRow(new Object[] {
            emp.name, prefs[0], prefs[1], prefs[2], prefs[3], prefs[4], prefs[5], prefs[6]
        });
    }

    private void removeSelected() {
        int row = getSelectedRow();
        if (row < 0) {
            JOptionPane.showMessageDialog(this, "Select an employee in the table to remove.");
            return;
        }
        employees.remove(row);
        tableModel.removeRow(row);
    }

    private void loadSampleData() {
        java.util.List<ScheduleManager.Employee> sample = ScheduleManager.collectInput();
        for (ScheduleManager.Employee e : sample) {
            if (!employees.stream().anyMatch(emp -> emp.name.equals(e.name))) {
                addEmployeeToTable(e);
            }
        }
        JOptionPane.showMessageDialog(this, "Sample employees added. You can add more or generate the schedule.");
    }

    private void clearAll() {
        int n = employees.size();
        employees.clear();
        for (int i = n - 1; i >= 0; i--) tableModel.removeRow(i);
    }

    private int getSelectedRow() {
        int r = employeeTable.getSelectedRow();
        return r >= 0 && r < employees.size() ? r : -1;
    }

    private void generateSchedule() {
        if (employees.isEmpty()) {
            JOptionPane.showMessageDialog(this, "Add at least one employee first.");
            return;
        }
        statusLabel.setText("Generating...");
        outputArea.setText("Generating schedule...");
        repaint();

        SwingWorker<String, Void> worker = new SwingWorker<String, Void>() {
            @Override
            protected String doInBackground() {
                return ScheduleManager.runAndGetScheduleOutput(employees);
            }
            @Override
            protected void done() {
                try {
                    String result = get();
                    outputArea.setText(result);
                    statusLabel.setText("Done.");
                } catch (Exception ex) {
                    outputArea.setText("Error: " + ex.getMessage());
                    statusLabel.setText("Error.");
                }
            }
        };
        worker.execute();
    }

    public static void launch() {
        try {
            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
        } catch (Exception ignored) {}
        EventQueue.invokeLater(() -> {
            ScheduleManagerGUI gui = new ScheduleManagerGUI();
            gui.setVisible(true);
        });
    }

    public static void main(String[] args) {
        launch();
    }

    // ---- Add Employee Dialog ----
    private static class AddEmployeeDialog extends JDialog {
        private final JTextField nameField;
        private final JComboBox<String>[][] dayCombos; // [day][0=1st, 1=2nd, 2=3rd]
        private ScheduleManager.Employee result;

        AddEmployeeDialog(Frame parent) {
            super(parent, "Add employee", true);
            setSize(520, 420);
            setLocationRelativeTo(parent);
            JPanel p = new JPanel(new BorderLayout(10, 10));
            p.setBorder(new EmptyBorder(12, 12, 12, 12));

            JPanel top = new JPanel(new FlowLayout(FlowLayout.LEFT));
            top.add(new JLabel("Name:"));
            nameField = new JTextField(20);
            top.add(nameField);
            p.add(top, BorderLayout.NORTH);

            JPanel grid = new JPanel(new GridLayout(8, 4, 4, 4));
            grid.add(new JLabel("Day"));
            grid.add(new JLabel("1st"));
            grid.add(new JLabel("2nd"));
            grid.add(new JLabel("3rd"));
            dayCombos = new JComboBox[7][3];
            for (int day = 0; day < 7; day++) {
                grid.add(new JLabel(DAYS[day]));
                for (int c = 0; c < 3; c++) {
                    JComboBox<String> cb = new JComboBox<>(SHIFTS);
                    cb.setSelectedIndex(c);
                    dayCombos[day][c] = cb;
                    grid.add(cb);
                }
            }
            p.add(grid, BorderLayout.CENTER);

            JPanel buttons = new JPanel(new FlowLayout(FlowLayout.RIGHT));
            JButton ok = new JButton("Save");
            ok.addActionListener(e -> onSave());
            JButton cancel = new JButton("Cancel");
            cancel.addActionListener(e -> dispose());
            buttons.add(ok);
            buttons.add(cancel);
            p.add(buttons, BorderLayout.SOUTH);

            setContentPane(p);
        }

        private void onSave() {
            String name = nameField.getText().trim();
            if (name.isEmpty()) {
                JOptionPane.showMessageDialog(this, "Please enter employee name.");
                return;
            }
            Map<Integer, String> preferences = new HashMap<>();
            Map<Integer, java.util.List<String>> priority = new HashMap<>();
            for (int day = 0; day < 7; day++) {
                String first = (String) dayCombos[day][0].getSelectedItem();
                String second = (String) dayCombos[day][1].getSelectedItem();
                String third = (String) dayCombos[day][2].getSelectedItem();
                if (first.equals(second) || first.equals(third) || second.equals(third)) {
                    JOptionPane.showMessageDialog(this, "1st, 2nd, and 3rd choice must be different for each day.");
                    return;
                }
                preferences.put(day, first);
                java.util.List<String> list = new ArrayList<>();
                list.add(first);
                list.add(second);
                list.add(third);
                priority.put(day, list);
            }
            result = new ScheduleManager.Employee(name, preferences, priority);
            dispose();
        }

        ScheduleManager.Employee getResult() {
            return result;
        }
    }
}

File: Assignment-4/python/schedule_manager_gui.py
#!/usr/bin/env python3
"""
Optional GUI for Employee Schedule Manager.
Input: employee names and shift preferences per day (entered by user in GUI).
Output: final weekly schedule in a readable format.
Uses tkinter (standard library).
"""
import tkinter as tk
from tkinter import ttk, messagebox, scrolledtext
import random
from schedule_manager import (
    DAYS,
    SHIFTS,
    Employee,
    build_schedule_structure,
    first_pass_assign,
    backfill_shifts,
)

MIN_EMPLOYEES_PER_SHIFT = 2
MAX_DAYS_PER_EMPLOYEE = 5
FIRST_PASS_MAX_DAYS = 4
MAX_PER_SHIFT = 4


def run_scheduler_from_gui(employees_data: list) -> str:
    """Build schedule from list of {name, preferences: {day: shift}, priority: {day: [shifts]} or None}."""
    employees = []
    for d in employees_data:
        emp = Employee(
            name=d["name"],
            preferences=d.get("preferences", {}),
            priority_ranking=d.get("priority"),
        )
        employees.append(emp)
    schedule = build_schedule_structure()
    use_priority = any(e.priority_ranking for e in employees)
    first_pass_assign(employees, schedule, use_priority=use_priority)
    backfill_shifts(employees, schedule)
    return format_schedule_text(schedule)


def format_schedule_text(schedule) -> str:
    lines = [
        "=" * 70,
        "                    FINAL EMPLOYEE SCHEDULE - WEEK",
        "=" * 70,
    ]
    for day in range(7):
        lines.append(f"\n  {DAYS[day].upper()}")
        lines.append("  " + "-" * 60)
        for shift in SHIFTS:
            names = schedule[day][shift]
            names_str = ", ".join(names) if names else "(none)"
            lines.append(f"    {shift.capitalize():10} : {names_str}")
    lines.append("\n" + "=" * 70)
    return "\n".join(lines)


class AddEmployeeDialog:
    """Dialog to enter employee name and shift preferences for each day."""

    def __init__(self, parent, on_save, edit_data=None):
        self.on_save = on_save
        self.edit_data = edit_data
        self.result = None
        self.win = tk.Toplevel(parent)
        self.win.title("Edit employee" if edit_data else "Add employee")
        self.win.transient(parent)
        self.win.grab_set()
        self.win.geometry("520x380")

        f = ttk.Frame(self.win, padding=10)
        f.pack(fill=tk.BOTH, expand=True)

        ttk.Label(f, text="Employee name:").grid(row=0, column=0, sticky=tk.W, pady=2)
        self.name_var = tk.StringVar(value=(edit_data["name"] if edit_data else ""))
        ttk.Entry(f, textvariable=self.name_var, width=25).grid(row=0, column=1, columnspan=2, sticky=tk.W, pady=2)

        ttk.Label(f, text="Preferred shift for each day (optional: 2nd and 3rd choice for priority):").grid(
            row=1, column=0, columnspan=4, sticky=tk.W, pady=(12, 4)
        )
        # Headers
        ttk.Label(f, text="Day").grid(row=2, column=0, padx=(0, 12), pady=2)
        ttk.Label(f, text="1st choice").grid(row=2, column=1, padx=4, pady=2)
        ttk.Label(f, text="2nd choice").grid(row=2, column=2, padx=4, pady=2)
        ttk.Label(f, text="3rd choice").grid(row=2, column=3, padx=4, pady=2)

        self.day_vars = []
        for day in range(7):
            row_vars = []
            ttk.Label(f, text=DAYS[day]).grid(row=3 + day, column=0, sticky=tk.W, padx=(0, 12), pady=2)
            for choice in range(3):
                var = tk.StringVar(value=SHIFTS[choice] if not edit_data else self._get_choice(edit_data, day, choice))
                row_vars.append(var)
                cb = ttk.Combobox(f, textvariable=var, values=SHIFTS, state="readonly", width=10)
                cb.grid(row=3 + day, column=1 + choice, padx=4, pady=2)
            self.day_vars.append(row_vars)

        btn_f = ttk.Frame(f)
        btn_f.grid(row=11, column=0, columnspan=4, pady=16)
        ttk.Button(btn_f, text="Save", command=self._save).pack(side=tk.LEFT, padx=4)
        ttk.Button(btn_f, text="Cancel", command=self.win.destroy).pack(side=tk.LEFT, padx=4)

        self.win.protocol("WM_DELETE_WINDOW", self.win.destroy)

    def _get_choice(self, data, day, choice):
        if data.get("priority") and day in data["priority"] and choice < len(data["priority"][day]):
            return data["priority"][day][choice]
        if data.get("preferences") and day in data["preferences"] and choice == 0:
            return data["preferences"][day]
        return SHIFTS[choice]

    def _save(self):
        name = self.name_var.get().strip()
        if not name:
            messagebox.showwarning("Missing name", "Please enter employee name.", parent=self.win)
            return
        preferences = {}
        priority = {}
        for day in range(7):
            choices = [self.day_vars[day][c].get().strip().lower() for c in range(3)]
            if not all(c in SHIFTS for c in choices):
                messagebox.showwarning("Invalid shift", "Use only: morning, afternoon, evening.", parent=self.win)
                return
            if len(set(choices)) != 3:
                messagebox.showwarning("Duplicate choice", "1st, 2nd, and 3rd choice must be different per day.", parent=self.win)
                return
            preferences[day] = choices[0]
            priority[day] = choices
        self.result = {"name": name, "preferences": preferences, "priority": priority}
        self.on_save(self.result, self.edit_data)
        self.win.destroy()


class ScheduleManagerApp:
    def __init__(self):
        self.root = tk.Tk()
        self.root.title("Employee Schedule Manager")
        self.root.geometry("740x560")
        self.employees_data = []
        self._build_ui()

    def _build_ui(self):
        # Input frame
        input_frame = ttk.LabelFrame(self.root, text="Employees & shift preferences (enter via Add employee)", padding=8)
        input_frame.pack(fill=tk.X, padx=8, pady=4)

        btn_f = ttk.Frame(input_frame)
        btn_f.pack(fill=tk.X)
        ttk.Button(btn_f, text="Add employee", command=self._add_employee).pack(side=tk.LEFT, padx=2)
        ttk.Button(btn_f, text="Edit selected", command=self._edit_employee).pack(side=tk.LEFT, padx=2)
        ttk.Button(btn_f, text="Remove selected", command=self._remove_employee).pack(side=tk.LEFT, padx=2)
        ttk.Button(btn_f, text="Load sample data", command=self._load_sample_data).pack(side=tk.LEFT, padx=2)
        ttk.Button(btn_f, text="Clear all", command=self._clear_employees).pack(side=tk.LEFT, padx=2)
        ttk.Button(btn_f, text="Generate schedule", command=self._generate).pack(side=tk.RIGHT, padx=2)

        self.listbox = tk.Listbox(input_frame, height=5, width=70)
        self.listbox.pack(fill=tk.X, pady=4)
        ttk.Label(input_frame, text="Double-click a name to edit. Add at least 2–3 employees for a valid schedule.").pack(anchor=tk.W)

        # Output frame
        output_frame = ttk.LabelFrame(self.root, text="Final schedule", padding=8)
        output_frame.pack(fill=tk.BOTH, expand=True, padx=8, pady=4)
        self.output_text = scrolledtext.ScrolledText(output_frame, wrap=tk.WORD, font=("Consolas", 10), height=18)
        self.output_text.pack(fill=tk.BOTH, expand=True)
        self.output_text.insert(tk.END, "Add employees (name + preferred shift per day), then click 'Generate schedule'.")
        self.output_text.config(state=tk.DISABLED)

        self.listbox.bind("<Double-Button-1>", lambda e: self._edit_employee())

    def _add_employee(self):
        def save(data, _):
            self.employees_data.append(data)
            self.listbox.insert(tk.END, data["name"])

        AddEmployeeDialog(self.root, save)

    def _edit_employee(self):
        sel = self.listbox.curselection()
        if not sel:
            messagebox.showinfo("Select employee", "Select an employee in the list to edit.")
            return
        idx = sel[0]
        edit_data = self.employees_data[idx]

        def save(data, original):
            self.employees_data[idx] = data
            self.listbox.delete(idx)
            self.listbox.insert(idx, data["name"])

        AddEmployeeDialog(self.root, save, edit_data=edit_data)

    def _remove_employee(self):
        sel = self.listbox.curselection()
        if not sel:
            messagebox.showinfo("Select employee", "Select an employee in the list to remove.")
            return
        idx = sel[0]
        del self.employees_data[idx]
        self.listbox.delete(idx)

    def _load_sample_data(self):
        from schedule_manager import collect_input
        for e in collect_input():
            if not any(d["name"] == e.name for d in self.employees_data):
                self.employees_data.append({
                    "name": e.name,
                    "preferences": e.preferences,
                    "priority": e.priority_ranking,
                })
                self.listbox.insert(tk.END, e.name)
        if self.employees_data:
            messagebox.showinfo("Sample data", "Sample employees added. You can edit or add more, then Generate schedule.")

    def _clear_employees(self):
        self.employees_data.clear()
        self.listbox.delete(0, tk.END)

    def _generate(self):
        if not self.employees_data:
            messagebox.showinfo("Info", "Add at least one employee first (Add employee or Load sample data).")
            return
        if len(self.employees_data) < 2:
            messagebox.showwarning("Few employees", "At least 2 employees are recommended so every shift can have 2 people.")
        random.seed(42)
        try:
            out = run_scheduler_from_gui(self.employees_data)
        except Exception as ex:
            messagebox.showerror("Error", str(ex))
            return
        self.output_text.config(state=tk.NORMAL)
        self.output_text.delete(1.0, tk.END)
        self.output_text.insert(tk.END, out)
        self.output_text.config(state=tk.DISABLED)

    def run(self):
        self.root.mainloop()


if __name__ == "__main__":
    ScheduleManagerApp().run()
